velox/core/PlanNode.h

```diff
@@ -1449,4 +1449,102 @@ class AssignUniqueIdNode : public PlanNode {
   std::shared_ptr<std::atomic_int64_t> uniqueIdCounter_;
 };
 
+/// PlanNode used for evaluating Sql window functions.
+/// All window functions evaluated in the operator have the same
+/// window spec (partition keys + order columns).
+/// If no partition keys are specified, then all input rows
+/// are considered to be in a single partition.
+/// If no order by columns are specified, then the input rows order
+/// is non-deterministic.
+/// Each window function also has a frame which specifies the sliding
+/// window over which it is computed. The frame
+/// could be RANGE (based on peers which are all rows with the same
+/// ORDER BY value) or ROWS (position based).
+/// The frame bound types are CURRENT_ROW, (expression or UNBOUNDED)
+/// ROWS_PRECEDING and (expression or UNBOUNDED) ROWS_FOLLOWING.
+/// The WindowNode has one passthrough output column for each input
+/// column followed by the results of the window functions.
+class WindowNode : public PlanNode {
+ public:
+  enum class WindowType { kRange, kRows };
+
+  enum class BoundType {
+    kUnboundedPreceding,
+    kPreceding,
+    kCurrentRow,
+    kFollowing,
+    kUnboundedFollowing
+  };
+
+  struct Frame {
+    WindowType type;
+    BoundType startType;
+    FieldAccessTypedExprPtr startValue;
```

Comment:
I wonder if it makes sense to use a base class here to allow for constant values.

Reply:
I couldn't entirely follow this comment. Please can you elaborate more.

Reply:
I expect that most queries would have fixes constants for start and end values for the frame. In this case it might be helpful to specify these as ConstantTypedExpr as this may allow for optimizing the implementation. To allow both FieldAccessTypedExprPtr and ConstantTypedExprPtr, we'd need to change the type here to base class TypedExprPtr.